pytorch_finufft/functional.py

```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
+
+        finufftkwargs = {k: v for k, v in finufftkwargs.items()}
+        _mode_ordering = finufftkwargs.pop("modeord", 1)
+        _i_sign = finufftkwargs.pop("isign", -1)
+
+        # if fftshift:
+        #     if _mode_ordering != 1:  # This seems like it is the wrong way round???????
+        #         raise ValueError(
+        #             "Double specification of ordering; only one of fftshift and "
+        #             "modeord should be provided."
+        #         )
+        #     _mode_ordering = 0
+
+        ndim = points.shape[0]
+        if _mode_ordering == 1:
+            targets = torch.fft.fftshift(targets, dim=tuple(range(-ndim, 0)))
+
+
+        ctx.isign = _i_sign
+        ctx.mode_ordering = _mode_ordering
+        ctx.finufftkwargs = finufftkwargs
+
+        ctx.save_for_backward(points, targets)
+
+        nufft_func = get_nufft_func(ndim, 2, points.device.type)
+
+        finufft_out = nufft_func(
+            *points.data.numpy(),
+            targets.data.numpy(),
+            isign=_i_sign,
+            **finufftkwargs,
+        )
+
+        return finufft_out
+
+    @staticmethod
+    def backward(
+        ctx: Any, grad_output: torch.Tensor
+    ) -> Tuple[
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        None,
+        None,
+        None,
+    ]:
+        """
+        Implements derivatives wrt. each argument in the forward method.
+
+        Parameters
+        ----------
+        ctx : Any
+            Pytorch context object
+        grad_output : torch.Tensor
+            Backpass gradient output.
+
+        Returns
+        -------
+        Tuple[ Union[torch.Tensor, None], ...]
+            A tuple of derivatives wrt. each argument in the forward method
+        """
+        _i_sign = ctx.isign
+        _mode_ordering = ctx.mode_ordering
+        finufftkwargs = ctx.finufftkwargs
+
+        points, targets = ctx.saved_tensors
+        device = points.device
+
+        # start_points = -(np.array(targets.shape) // 2)
+        # end_points = start_points + targets.shape
+        # slices = tuple(slice(start, end) for start, end in zip(start_points, end_points))
+
+        # # CPU idiosyncracy that needs to be done differently
+        # k_ramps = torch.from_numpy(np.mgrid[slices], dtype=points.dtype)
+
+        grad_points = grad_targets = None
+        ndim = points.shape[0]
+
+
+        if ctx.needs_input_grad[0]:
+            # wrt points
+            start_points = -(torch.tensor(targets.shape, device=device) // 2)
+            end_points = start_points + torch.tensor(targets.shape, device=device)
+            coord_ramps = torch.stack(
+                torch.meshgrid(
+                    *(
+                        torch.arange(start, end, device=device)
+                        for start, end in zip(start_points, end_points)
+                    ),
+                    indexing="ij",
+                )
+            )
```

Comment:
[optional] if this is exactly the same code as type1, I wonder if it is worth pulling out into a helper function

Reply:
good idea

Reply:
done